### ISSUE TITLE
ApmMiddleware.InvokeAsync() invoked multiple times

### ISSUE THREAD

**Issue by kanadaj** [elastic/apm-agent-dotnet#1019]
**Describe the bug**
In error stacktraces, the ApmMiddleware seems to appear multiple times in the stacktrace in Sentry:
```
  Module "Elastic.Apm.AspNetCore.ApmMiddleware", in InvokeAsync
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "Elastic.Apm.AspNetCore.ApmMiddleware", in InvokeAsync
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
  Module "Elastic.Apm.AspNetCore.ApmMiddleware", in InvokeAsync
  Module "System.Runtime.ExceptionServices.ExceptionDispatchInfo", in Throw
```

or without Sentry:
```
   at Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
   at Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
   at Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
```

This suggests that the middleware is added to the pipeline multiple times and I'm unsure why. 

The code to add the middleware is as follows:

```cs
            app.UseElasticApm(Configuration, new HttpDiagnosticsSubscriber(), new AspNetCoreDiagnosticsSubscriber(), new EfCoreDiagnosticsSubscriber(), new ElasticsearchDiagnosticsSubscriber());
```

The middlewares after ElasticApm seem to work as expected so I'm wondering what exactly could be happening here. 

I was wondering if it could be maybe async/await monkey business, but the stacktrace has no indication of any of the async callbacks/iterators I'd expect to see if that was the case. It just strikes me as weird.

Using ElasticApm 1.6.1

**Reproduction steps**

1. Create a new ASP.NET Core project
2. Add Elastic APM via NuGet
3. Add ElasticApm to the pipeline
4. Trigger any exception
5. Check stacktrace for ApmMiddleware invocations

**Expected behavior**
Assuming the middleware doesn't recursively invoke itself, I'd expect only one call of InvokeAsync there. I can't see any indication of UseElasticApm adding the Middleware multiple times.

**Comment by russcam** [comment]
Hi @kanadaj,

I've out together a sample app based on the reproduction steps: https://github.com/russcam/ApmExample
I **can't replicate** the issue however; the error stacktrace for the `/exception` endpoint is

```
boom!
ApmExample.Startup+<>c+<<Configure>b__5_2>d in <Configure>b__5_2 at line 55
Microsoft.AspNetCore.Routing.EndpointMiddleware+<<Invoke>g__AwaitRequestTask|6_0>d in <Invoke>g__AwaitRequestTask|6_0
Elastic.Apm.AspNetCore.ApmMiddleware+<InvokeAsync>d__4 in InvokeAsync
```

Do you have other middleware registered?

**Comment by kanadaj** [comment]
I've created a repro: https://github.com/kanadaj/ElasticApmMultipleInvocationRepro

The issue triggers on both `/exception` (MVC route) and `/route-exception` (Directly mapped with `MapGet()`). Otherwise this is pretty much the default MVC pipeline you get. I'm using dotnet SDK 5.0 but this happened even before I updated my SDK.

**Comment by russcam** [comment]
Thanks for the repro @kanadaj, it's been really helpful for investigation.

Going through the flow when hitting the `/exception` endpoint in the example, I don't believe there to be an issue here. The `DeveloperExceptionPageMiddleware` or other middleware that collects stacktrace frames after Elastic `ApmMiddleware` will show a frame for the awaited tasks inside `ApmMiddleware.InvokeAsync()`. For example, here's the three frames that relate to `ApmMiddleware.InvokeAsync()`  seen on the developer exception page

> NotImplementedException: The method or operation is not implemented.
> ElasticApmMultipleInvocationRepro.Controllers.HomeController.Exception() in HomeController.cs
> +
>             throw new NotImplementedException();
> lambda_method(Closure , object , object[] )
> Microsoft.Extensions.Internal.ObjectMethodExecutor.Execute(object target, object[] parameters)
> Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor+SyncActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, object controller, object[] arguments)
> Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
> Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
> Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
> Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
> Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
> Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, object state, bool isCompleted)
> Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
> Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
> Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
> Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
> Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
> Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
> Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
> Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
> Elastic.Apm.AspNetCore.ApmMiddleware.InvokeAsync(HttpContext context)
> Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

![developer exception page](https://user-images.githubusercontent.com/208231/99338204-774fec80-28cf-11eb-9882-6962efdc656c.png)

![stacktrace frame information collected for developer exception page](https://user-images.githubusercontent.com/208231/99338058-3c4db900-28cf-11eb-8ebb-0fcbefe95bae.png)

The IL offsets indicate that these frames relate to different parts of `InvokeAsync()`, which look to be the following awaited tasks

https://github.com/elastic/apm-agent-dotnet/blob/739e2a2cae99fa86ad987aedf3af0c740a8b5872/src/Elastic.Apm.AspNetCore/ApmMiddleware.cs#L53

https://github.com/elastic/apm-agent-dotnet/blob/739e2a2cae99fa86ad987aedf3af0c740a8b5872/src/Elastic.Apm.AspNetCore/ApmMiddleware.cs#L57

and 

https://github.com/elastic/apm-agent-dotnet/blob/739e2a2cae99fa86ad987aedf3af0c740a8b5872/src/Elastic.Apm.AspNetCore/ApmMiddleware.cs#L65

In summary, this doesn't look like a bug.

**Comment by kanadaj** [comment]
Oh alright, just async shenanigans. Had to check anyway since it looked mighty suspicious in the stacktrace :)

**Comment by russcam** [comment]
No worries, thanks for raising!
